### ISSUE TITLE
[Proposal] Add support for DPI scaling

### ISSUE THREAD

**Issue by wesbluemarine** [Barrett17/V-OS#27]
Font and widgets scaling should be a must for a good user experience.

**Comment by Barrett17** [comment]
It would be useful to have mockups/showcases for this stuff :)

**Comment by carmiac** [comment]
I'm curious about this one. For the most part, the widgets already support scaling, though the way to change the UI scale is very non-obvious.

The widgets that I know of that don't already support scaling properly are:

- ColumnListView arrow
- Tracker list view arrow
- Menu shortcut key icons
- Sliders (which are a bit of a mess in general)
- Replicant handle

I have code to fix the first 4, but the 5th is a much harder problem. I also have code that tweaks the Appearance preflet to explain the scaling system and adds a single slider that can sett all of the widget/font sizes at once. Does that sound interesting and cover this one?

**Comment by Numerio** [comment]
Yes indeed we'd love to see it. The general broader discussion was in fact if we wanted to switch to normal scaling rather than scaling with fonts size which is still an open question, I'd love to hear what you think about it.

**Comment by carmiac** [comment]
Ok, great! I'll work on preping the PRs.

For overall scaling I think there is some good, bad and ugly with the current set up. The bad is mostly easy UX fixes, the ugly will take some thought.

Good:
Using the font size as a scaling base seems like an odd choice at first, but actually makes a decent amount of sense in practice. When widgets properly implement it, it should prevent many of the layout/font issues that other systems see when using a custom font that may have an unexpected size. e.g. Labels overflowing their boxes because not all fonts are the same pixel size even when they are the same pt size. I also like that it allows for fine tuning the UI scale. It's nice for high res monitors and my tired old eyes.

Bad:
Different widgets scale off of different fonts! Most scale off of the plain font, but some scale off of the bold or menu font sizes and IIRC there are even some that use the monospaced size. From a user perspective this can be a bit weird. More importantly there isn't an obvious link between setting the font size and the overall UI scale mentioned anywhere, and the need to restart things really obscures it. That's an easy enough fix in the Appearance preflet though.

The other problem is that there are several other scaling things that don't use the font sizes at all: Desktop and Tracker Icons are the first that come to mind, but there are probably others. Again not a huge issue, just a few more knobs to set during initial install. 

It might be worth having a "Scaling" tab in the Appearance preflet with an overall size slider that sets everything at once, and then individual sliders for everything else that can override the main one. And then leave the existing size settings where they are for people who really want to tweak things.

Ugly:
Multimonitor and per-Workspace resolution support. This is where it falls apart. If you set things to one size for a given resolution and then move a window to a monitor/workspace with a different DPI, everything is either way too large or way too small. Qt handles this with a per-device pixel ratio that it queries from the OS IIRC. I think something like that might work well. If the pixel ratio is set to 1 for the primary device, and then adjusted it per device based on either a queried DPI or user setting, app_server could then scale everything from that ratio and it should just work for existing code. There are probably some tricky details to work out that I'm not aware of.

**Comment by Numerio** [comment]
I just don't understand what's wrong with the scaling like it work on linux. On my 4k screen I just set 1.5 or 1.75 scaling and it looks good to me (99% of stuff seems working). I mean, I get the idea of using fonts, maybe in an ideal world would work, I remember when they decided to implement it like that, a lot of people in the ml didn't agree. But I have never really looked in deep to take an informed decision. However, happy to receive patches, also the UX side of the interface kit is currently an area where I just don't have the time to work on, so if you want to help improving the new preflets and whatnot you're welcome. BTW join in the telegram channel if you aren't already.

**Comment by carmiac** [comment]
Just to clarify a bit, basing it off font size wouldn't be my first choice, and I'm unfamiliar with the discussion that lead to the decision. 

But it makes a good enough second choice that I'd rather get the few remaining widgets that don't scale right working correctly before changing everything to a different system that will require at least digging into app_server's drawing system and probably updating every control look and widget drawing function. 

I'll do a longer explanation later, with code examples which I think will explain it better.

**Comment by carmiac** [comment]
And I'm in the Telegram, I just don't like chat rooms for most things.
